File: sdk/java/WePay.java
import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.security.*;
import java.security.spec.PKCS8EncodedKeySpec;
import java.util.*;
import javax.crypto.Cipher;
import com.google.gson.Gson;
import com.google.gson.JsonObject;
import org.apache.commons.codec.binary.Base64;
import org.apache.commons.codec.digest.DigestUtils;

/**
 * WePay Java SDK
 * 支持 MD5 和 RSA 签名的支付网关 SDK
 */
public class WePay {
    private String mchId;
    private String mchKey;
    private String baseUrl;
    private String signType;
    private String privateKeyPem;
    private Gson gson = new Gson();

    /**
     * 初始化 SDK
     */
    public WePay(String mchId, String mchKey, String baseUrl, String signType, String privateKeyPem) {
        this.mchId = mchId;
        this.mchKey = mchKey;
        this.baseUrl = baseUrl.replaceAll("/$", "");
        this.signType = signType.toUpperCase();
        this.privateKeyPem = privateKeyPem;
    }

    /**
     * 生成签名
     */
    public String sign(Map<String, String> params) throws Exception {
        if ("RSA".equals(signType)) {
            return signRsa(params);
        } else {
            return signMd5(params);
        }
    }

    /**
     * MD5 签名
     */
    private String signMd5(Map<String, String> params) {
        List<String> keys = new ArrayList<>(params.keySet());
        Collections.sort(keys);

        StringBuilder sb = new StringBuilder();
        for (String key : keys) {
            String value = params.get(key);
            if (!"sign".equals(key) && !"sign_type".equals(key) && value != null && !value.isEmpty()) {
                sb.append(key).append("=").append(value).append("&");
            }
        }
        sb.deleteCharAt(sb.length() - 1);
        sb.append(mchKey);

        return DigestUtils.md5Hex(sb.toString());
    }

    /**
     * RSA-SHA256 签名
     */
    private String signRsa(Map<String, String> params) throws Exception {
        if (privateKeyPem == null || privateKeyPem.isEmpty()) {
            throw new Exception("RSA 签名需要提供私钥");
        }

        List<String> keys = new ArrayList<>(params.keySet());
        Collections.sort(keys);

        StringBuilder sb = new StringBuilder();
        for (String key : keys) {
            String value = params.get(key);
            if (!"sign".equals(key) && !"sign_type".equals(key) && value != null && !value.isEmpty()) {
                sb.append(key).append("=").append(value).append("&");
            }
        }
        sb.deleteCharAt(sb.length() - 1);

        PrivateKey privateKey = loadPrivateKey(privateKeyPem);
        Signature signature = Signature.getInstance("SHA256withRSA");
        signature.initSign(privateKey);
        signature.update(sb.toString().getBytes(StandardCharsets.UTF_8));

        byte[] signBytes = signature.sign();
        return Base64.encodeBase64String(signBytes);
    }

    /**
     * 加载 RSA 私钥
     */
    private PrivateKey loadPrivateKey(String privateKeyPem) throws Exception {
        String privateKeyContent = privateKeyPem
            .replace("-----BEGIN PRIVATE KEY-----", "")
            .replace("-----END PRIVATE KEY-----", "")
            .replaceAll("\\s", "");

        byte[] decodedKey = Base64.decodeBase64(privateKeyContent);
        PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(decodedKey);
        KeyFactory keyFactory = KeyFactory.getInstance("RSA");
        return keyFactory.generatePrivate(keySpec);
    }

    /**
     * 统一下单
     */
    public JsonObject createOrder(Map<String, String> params) throws Exception {
        params.put("mch_id", mchId);
        params.put("sign_type", signType);
        params.put("sign", sign(params));

        return post("/gateway/create", params);
    }

    /**
     * 查询订单
     */
    public JsonObject queryOrder(String outTradeNo) throws Exception {
        Map<String, String> params = new HashMap<>();
        params.put("mch_id", mchId);
        params.put("out_trade_no", outTradeNo);
        params.put("sign_type", signType);
        params.put("sign", sign(params));

        return post("/gateway/query", params);
    }

    /**
     * 申请退款
     */
    public JsonObject refund(String tradeNo, String refundAmount) throws Exception {
        Map<String, String> params = new HashMap<>();
        params.put("mch_id", mchId);
        params.put("trade_no", tradeNo);
        params.put("refund_amount", refundAmount);
        params.put("sign_type", signType);
        params.put("sign", sign(params));

        return post("/gateway/refund", params);
    }

    /**
     * 验证异步通知签名
     */
    public boolean verifyNotify(Map<String, String> params, String sign) throws Exception {
        String calcSign = this.sign(params);
        return calcSign.equals(sign);
    }

    /**
     * POST 请求
     */
    private JsonObject post(String path, Map<String, String> params) throws Exception {
        URL url = new URL(baseUrl + path);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("POST");
        conn.setRequestProperty("Content-Type", "application/json");
        conn.setDoOutput(true);
        conn.setConnectTimeout(10000);

        String jsonBody = gson.toJson(params);
        try (OutputStream os = conn.getOutputStream()) {
            os.write(jsonBody.getBytes(StandardCharsets.UTF_8));
        }

        int responseCode = conn.getResponseCode();
        if (responseCode != 200) {
            throw new Exception("HTTP 错误: " + responseCode);
        }

        StringBuilder response = new StringBuilder();
        try (BufferedReader br = new BufferedReader(new InputStreamReader(conn.getInputStream(), StandardCharsets.UTF_8))) {
            String line;
            while ((line = br.readLine()) != null) {
                response.append(line);
            }
        }

        return gson.fromJson(response.toString(), JsonObject.class);
    }

    /**
     * 示例代码
     */
    public static void main(String[] args) {
        try {
            WePay wepay = new WePay(
                "M515637",
                "nca3twhvpqveixu2hdeutb6utpiet6k7",
                "http://127.0.0.1:8088",
                "MD5",
                ""
            );

            Map<String, String> orderParams = new HashMap<>();
            orderParams.put("out_trade_no", "order_java_" + System.currentTimeMillis());
            orderParams.put("pay_type", "wxpay");
            orderParams.put("amount", "0.01");
            orderParams.put("subject", "Java SDK 测试订单");

            JsonObject result = wepay.createOrder(orderParams);
            System.out.println("下单结果:");
            System.out.println(gson.toJson(result));
        } catch (Exception e) {
            System.err.println("错误: " + e.getMessage());
            e.printStackTrace();
        }
    }
}

File: sdk/go/wepay.go
package wepay

import (
	"bytes"
	"crypto"
	"crypto/md5"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"encoding/pem"
	"fmt"
	"io"
	"net/http"
	"sort"
	"strings"
	"time"
)

// WePay SDK 结构体
type WePay struct {
	MchID       string
	MchKey      string
	BaseURL     string
	SignType    string
	PrivateKey  *rsa.PrivateKey
	HTTPClient  *http.Client
}

// 创建新的 WePay SDK 实例
func New(mchID, mchKey, baseURL, signType, privateKeyPEM string) (*WePay, error) {
	w := &WePay{
		MchID:      mchID,
		MchKey:     mchKey,
		BaseURL:    strings.TrimRight(baseURL, "/"),
		SignType:   strings.ToUpper(signType),
		HTTPClient: &http.Client{Timeout: 10 * time.Second},
	}

	if w.SignType == "RSA" && privateKeyPEM != "" {
		privateKey, err := loadPrivateKey(privateKeyPEM)
		if err != nil {
			return nil, err
		}
		w.PrivateKey = privateKey
	}

	return w, nil
}

// 加载 RSA 私钥
func loadPrivateKey(privateKeyPEM string) (*rsa.PrivateKey, error) {
	block, _ := pem.Decode([]byte(privateKeyPEM))
	if block == nil {
		return nil, fmt.Errorf("私钥格式错误")
	}

	privateKey, err := x509.ParsePKCS8PrivateKey(block.Bytes)
	if err != nil {
		return nil, err
	}

	rsaKey, ok := privateKey.(*rsa.PrivateKey)
	if !ok {
		return nil, fmt.Errorf("不是 RSA 私钥")
	}

	return rsaKey, nil
}

// 生成签名
func (w *WePay) Sign(params map[string]string) (string, error) {
	if w.SignType == "RSA" {
		return w.signRSA(params)
	}
	return w.signMD5(params), nil
}

// MD5 签名
func (w *WePay) signMD5(params map[string]string) string {
	// 排序参数
	keys := make([]string, 0, len(params))
	for k := range params {
		if k != "sign" && k != "sign_type" && params[k] != "" {
			keys = append(keys, k)
		}
	}
	sort.Strings(keys)

	// 构建签名字符串
	var sb strings.Builder
	for i, k := range keys {
		if i > 0 {
			sb.WriteString("&")
		}
		sb.WriteString(k)
		sb.WriteString("=")
		sb.WriteString(params[k])
	}
	sb.WriteString(w.MchKey)

	// MD5 签名
	hash := md5.Sum([]byte(sb.String()))
	return fmt.Sprintf("%x", hash)
}

// RSA-SHA256 签名
func (w *WePay) signRSA(params map[string]string) (string, error) {
	if w.PrivateKey == nil {
		return "", fmt.Errorf("RSA 签名需要提供私钥")
	}

	// 排序参数
	keys := make([]string, 0, len(params))
	for k := range params {
		if k != "sign" && k != "sign_type" && params[k] != "" {
			keys = append(keys, k)
		}
	}
	sort.Strings(keys)

	// 构建签名字符串
	var sb strings.Builder
	for i, k := range keys {
		if i > 0 {
			sb.WriteString("&")
		}
		sb.WriteString(k)
		sb.WriteString("=")
		sb.WriteString(params[k])
	}

	// RSA-SHA256 签名
	hash := sha256.Sum256([]byte(sb.String()))
	signature, err := rsa.SignPKCS1v15(rand.Reader, w.PrivateKey, crypto.SHA256, hash[:])
	if err != nil {
		return "", err
	}

	return base64.StdEncoding.EncodeToString(signature), nil
}

// 统一下单
func (w *WePay) CreateOrder(params map[string]string) (map[string]interface{}, error) {
	params["mch_id"] = w.MchID
	params["sign_type"] = w.SignType

	sign, err := w.Sign(params)
	if err != nil {
		return nil, err
	}
	params["sign"] = sign

	return w.post("/gateway/create", params)
}

// 查询订单
func (w *WePay) QueryOrder(outTradeNo string) (map[string]interface{}, error) {
	params := map[string]string{
		"mch_id":       w.MchID,
		"out_trade_no": outTradeNo,
		"sign_type":    w.SignType,
	}

	sign, err := w.Sign(params)
	if err != nil {
		return nil, err
	}
	params["sign"] = sign

	return w.post("/gateway/query", params)
}

// 申请退款
func (w *WePay) Refund(tradeNo, refundAmount string) (map[string]interface{}, error) {
	params := map[string]string{
		"mch_id":         w.MchID,
		"trade_no":       tradeNo,
		"refund_amount":  refundAmount,
		"sign_type":      w.SignType,
	}

	sign, err := w.Sign(params)
	if err != nil {
		return nil, err
	}
	params["sign"] = sign

	return w.post("/gateway/refund", params)
}

// 验证异步通知签名
func (w *WePay) VerifyNotify(params map[string]string, sign string) (bool, error) {
	calcSign, err := w.Sign(params)
	if err != nil {
		return false, err
	}
	return calcSign == sign, nil
}

// POST 请求
func (w *WePay) post(path string, params map[string]string) (map[string]interface{}, error) {
	url := w.BaseURL + path

	jsonData, err := json.Marshal(params)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/json")

	resp, err := w.HTTPClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("HTTP 错误: %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var result map[string]interface{}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}

	return result, nil
}
